src/Microsoft.Azure.WebJobs.Host/Listeners/FunctionListener.cs

```diff
@@ -114,6 +114,17 @@ public void Dispose()
                         message = $"Listener successfully started for function '{_descriptor.ShortName}' after {attempt} retries.";
                         _trace.Info(message);
                         _logger?.LogInformation(message);
+
+                        if (cancellationToken.IsCancellationRequested)
+                        {
+                            // stop has been called while we were in the process of starting
+                            // so we need to stop this listener
+                            await StopAsync(cancellationToken);
```

Comment:
Now there's a chance that Stop can be called multiple times on a single listener. Do we know how different listeners will react to that?

Reply:
You're talking about the case where the retry succeeds to start the listener, sets _started to true, context switches to an outer StopAsync call which will cancel the token and call Stop, context switch back and the retry loop sees the cancellation and calls stop again? Yes, there is a race there I'll need to synchronize.

Reply:
Ok, I've added synchronization on StopAsync

Reply:
Looks good!